src/cryptography/hazmat/bindings/openssl/binding.py

```diff
@@ -176,9 +188,11 @@ def _ensure_ffi_initialized(cls) -> None:
                     cls._legacy_provider = cls.lib.OSSL_PROVIDER_load(
                         cls.ffi.NULL, b"legacy"
                     )
-                    _openssl_assert(
-                        cls.lib, cls._legacy_provider != cls.ffi.NULL
-                    )
+                    if cls._legacy_provider != cls.ffi.NULL:
+                        cls._legacy_provider_loaded = True
+                    else:
+                        _legacy_provider_error()
```

Comment:
This still loads the legacy provider if you set CRYPTOGRAPHY_OPENSSL_NO_LEGACY, which we should not do.

We need to check the env var and _then_ load it.